db/collection.go

```diff
@@ -1082,6 +1088,108 @@ func (c *collection) save(
 	return headNode.Cid(), nil
 }
 
+func (c *collection) validateOneToOneLinkDoesntAlreadyExist(
+	ctx context.Context,
+	txn datastore.Txn,
+	docKey string,
+	fieldDescription client.FieldDescription,
+	value any,
+) error {
+	if !fieldDescription.RelationType.IsSet(client.Relation_Type_INTERNAL_ID) {
+		return nil
+	}
+
+	if value == nil {
+		return nil
+	}
+
+	objFieldDescription, ok := c.desc.Schema.GetField(strings.TrimSuffix(fieldDescription.Name, request.RelatedObjectID))
+	if !ok {
+		return client.NewErrFieldNotExist(strings.TrimSuffix(fieldDescription.Name, request.RelatedObjectID))
+	}
+	if !objFieldDescription.RelationType.IsSet(client.Relation_Type_ONEONE) {
+		return nil
+	}
+
+	fetcher := c.newFetcher()
+
+	err := fetcher.Init(ctx, txn, &c.desc, []client.FieldDescription{fieldDescription}, nil, nil, false, false)
```

Comment:
question: Wouldn't it help to use a filter here instead of having to filter bellow? This way you would only need to check if it's self and if otherwise documents are returned, a relationship has already been set.

Reply:
Yes, but a filter requires a mapping, which would need to be constructed, and then a manually constructed filter. Atm it really isn't worth that hassle I think as I think it would be uglier, and I'm doubtful it would have any meaningful impact on performance (atm) either.

Maybe in the future.

Reply:
cc: @AndrewSisley @fredcarle 

There is currently, and will continue to be *huge* savings both in time and space to use the Fetcher with the filter, and let the fetcher figure out how to effeciently give you the results.

In its current form, it will drastically reduce bytes copied, and in its future form it will drastically reduce the time to find all the docs matching the given filter.

At no point should any caller of the fetcher be doing its own filtering / field selection, unless it needs to merge/join with data from another collection/fetcher.

Reply:
I find that the fact that we have to do new,  Init, Start and FetchNext here is quite suboptimal. The field selection is done on Init so the other problem is the filtering that is happening within this function.  This will absolutely need to be refactored at some point in the near future but given that we don't have a friendly way to get with filter, I think this is acceptable for this PR.

Reply:
> I find that the fact that we have to do new, Init, Start and FetchNext here is quite suboptimal.

The Fetcher is a fairly substantial system, and as such, it needs to fit a lot of different use cases, and functionality. Which is why things are seperated into `Init`, `Start`. The `FetchNext` is irrelvant, as that would be the case regardless of how the fetcher would be constructed. And the `newFetcher` stuff is just a shim so we can redirect which type of fetcher we're going to use. So really, it just comes down to `Init` and `Start`, which is seperated, so we can do proper multi-range fetches without rebuilding the fetcher.

> it really isn't worth that hassle

I find this somewhat dubious, as this is a wildly ineffecient approach for starters (doing a full table scan, and collecting every single field of every single document, just to determine we're not duplicating one-one links), so the hassle here would result in drastically reducing both the time and space complexity of this approach (only getting the field we care about, the relation id field, and only returning the docs that actually match the filter, and seeking past the rest, where `Seek` is magnitudes faster than a `Get`)

> but given that we don't have a friendly way to get with filter
This is a function of the current mapping system, and at the moment we have to live with it. If we're really promoting a much less effecient and out right wasteful approach, simply cus we don't want to touch the mapper, then we need to shift priorities to address that.

For reference, the Type Join code in the planner does exactly this, constructing a new filter, from the mapper, and passes that to the Fetcher, so it can let the fetcher do the hard work of effeciently seeking and filters docs.
```go
// line 347 of type_join.go
fkIndex := &mapper.PropertyIndex{
    Index: n.subType.DocumentMap().FirstIndexOfName(n.subTypeFieldName + "_id"),
}
filter := map[connor.FilterKey]any{
    fkIndex: doc.GetKey(),
}
```

Additionally, there is a func in the `mapper` package that makes this slightly less annoying, producing:
```go
rawFilter := map[string]any{
    relatedFieldName + "_id": map[string]any{"_eq": dockey}
}

filter := mapper.ToFilter(immutable.Some(request.Filter{filter}), mapping)
```

Both are a hassle, and we can put whatever syntactic surgar in between. But choosing to use the Fetcher in an inefficient way to avoid having to write an addition ~6 lines, and instead manually writing ~4 lines to do your own filter check and spend more time and space, seems... wrong.

--
As [mentioned](https://github.com/sourcenetwork/defradb/pull/1500#discussion_r1192614605) in the Fetcher refactor PR, the ugliness in constructing the filter is based on the necessity of the filter needing the doc mapper.

Reply:
If you're happy with doing the whole fetcher manipulation here in the long run then it's really just adding the filter. The field selection is already passed to Init so not collecting every single field.

At the moment it has to be a whole table scan. We don't have indexes on the relational fields.

It would be nice to have a convenience function wrapping new, init and start but that would only really be useful if we're going to have multiple other locations that work directly with the fetcher like this.